### PR TITLE
Extending best practice guide for DFT

### PR DESCRIPTION
**Pull request by mgrub**
@eichstaedtPTB : Please check, if this extension of the "DFT and iDFT with PyDynamic - best practice guide and use cases" jupyter notebook fits your needs.

Meanwhile the notebook is included in the docs as well. Find the new version here:

- https://pydynamic.readthedocs.io/en/enhance_dft_example/examples/DFT%20and%20iDFT%20with%20PyDynamic%20-%20best%20practice%20guide%20and%20use%20cases.html

and the previous version here:

- https://pydynamic.readthedocs.io/en/latest/examples/DFT%20and%20iDFT%20with%20PyDynamic%20-%20best%20practice%20guide%20and%20use%20cases.html

Now we got the docs (and tests by the way) for the new public functions ready as well. Check it out: https://pydynamic.readthedocs.io/en/enhance_dft_example/PyDynamic.misc.html#PyDynamic.misc.tools.real_imag_2_complex

### CODE REVIEW

**Comment by BjoernLudwigPTB** [comment]
Is all that's needed really a code review and the corresponding refinements? Then we should resurrect that PR and get his done! I could take a look at it as well...

**Comment by BjoernLudwigPTB** [comment]
Unfortunately we have been rate-limited by CircleCI for the rest of the month, so we have to run the tests locally. I reconfigured the pipeline in #261 to avoid that in the future.

**Review by BjoernLudwigPTB — APPROVED** [review]
That's a very neat and useful display of PyDynamic's DTF capabilities. Thanks for that effort.

**Comment by BjoernLudwigPTB** [comment]
@mgrub from my side this is fine, are you ready to merge this?

File: test/test_tools.py [hunk 1]
```diff
@@ -96,3 +107,256 @@ def _minimum_is_zero(array: np.ndarray) -> bool:
...
+
+@given(
+    hnp.arrays(
+        dtype=hst.one_of(
+            hnp.unsigned_integer_dtypes(), hnp.integer_dtypes(), hnp.floating_dtypes()
+        ),
+        shape=hst.builds(
+            lambda shape_tuple: (shape_tuple[0], shape_tuple[1] * 2),
+            hnp.array_shapes(min_dims=2, max_dims=2, max_side=10),
+        ),
```
Comment (mgrub) [inline test/test_tools.py:279]: I think it would be useful to extract this into its own strategy.

File: test/test_tools.py [hunk 2]
```diff
@@ -96,3 +107,256 @@ def _minimum_is_zero(array: np.ndarray) -> bool:
...
+
+@given(hnp.arrays(dtype=hnp.scalar_dtypes(), shape=hypothesis_odd_dimension()))
+def test_real_imag_2_complex_vector_wrong_len(vector):
+    with pytest.raises(
+        ValueError,
+        match="separate_real_imag_of_vector: vector of real and imaginary "
+        "parts is "
+        "expected to contain exactly as many real as imaginary parts "
+        "but is of "
+        r"odd length=.*",
```
Comment (mgrub) [inline test/test_tools.py:341]: A small reminder :-)

**Review by mgrub — COMMENTED** [review]
This was a very good review and the implemented functionality goes (in a good sense) well beyond the initial intention of the pull request :-) 

As suggested in the comments, there are some minor tweaks remaining, but overall: great work 👍

File: src/PyDynamic/misc/tools.py [hunk 3]
```diff
@@ -21,6 +21,14 @@
...
   uncertainties ux
 * :func:`trimOrPad`: trim or pad (with zeros) a vector to desired length
+* :func:`real_imag_2_complex`: Take a np.ndarray with real and imaginary parts
+  and return dtype complex ndarray
+* :func:`separate_real_imag_of_mc_samples`: Split a np.ndarray containing MonteCarlo
+  samples real and imaginary parts
+* :func:`separate_real_imag_of_vector`: Split a np.ndarray containing real and
+  imaginary parts into half
+* :func:`complex_2_real_imag`: Take a np.ndarray with dtype complex and return
+  real and imaginary parts
```
[the remark this answers is not in the document]
Comment (BjoernLudwigPTB) [reply src/PyDynamic/misc/tools.py:31]: Good suggestion. Moved up in 24099c09.

File: src/PyDynamic/misc/tools.py [hunk 4]
```diff
@@ -520,3 +532,113 @@ def normalize_vector_or_matrix(numbers: np.ndarray) -> np.ndarray:
...
     return (numbers - translator) / normalizer
+
+
+def real_imag_2_complex(array: np.ndarray) -> np.ndarray:
+    r"""Take a np.ndarray with real and imaginary parts and return dtype complex ndarray
+
+    The input array :math:`x \in \mathbb R^n` representing a complex vector
+    :math:`y \in \mathbb C^{n/2}` has the form of the expected input of
+    some of the functions in the modules :mod:`PyDynamic.uncertainty.propagate_DFT` and
+    :mod:`PyDynamic.model_estimation.fit_filter`: :math:`x = \left( \operatorname{Re}(
```
[the remark this answers is not in the document]
Comment (BjoernLudwigPTB) [reply src/PyDynamic/misc/tools.py]: Resolved in ac2c6725.

File: test/conftest.py [hunk 5]
```diff
@@ -301,7 +324,7 @@ def hypothesis_dimension(
 ) -> int:
     minimum_dimension = min_value if min_value is not None else 1
     maximum_dimension = max_value if max_value is not None else 20
```
[the remark this answers is not in the document]
Comment (BjoernLudwigPTB) [reply test/conftest.py:326]: Resolved by ac2c6725.

File: test/conftest.py [hunk 6]
```diff
@@ -52,6 +52,29 @@ def hypothesis_reasonable_dimension_strategy(
     return draw(hst.integers(min_value=min_value, max_value=max_value))
```
[the remark this answers is not in the document]
Comment (BjoernLudwigPTB) [reply test/conftest.py:52]: Resolved by 7bf7dbd2.

File: test/conftest.py [hunk 7]
```diff
@@ -310,6 +333,72 @@ def hypothesis_dimension(
...
+
+
+@composite
+def hypothesis_even_dimension(
+    draw: Callable,
+    min_value: Optional[int] = None,
+    max_value: Optional[int] = None,
+) -> int:
+    minimum_dimension = min_value if min_value is not None else 2
+    maximum_dimension = max_value if max_value is not None else 20
```
[the remark this answers is not in the document]
Comment (BjoernLudwigPTB) [reply test/conftest.py:347]: Resolved by 7bf7dbd2.

File: test/conftest.py [hunk 8]
```diff
@@ -310,6 +333,72 @@ def hypothesis_dimension(
...
+        else draw(
+            hypothesis_even_dimension_strategy(
+                min_value=minimum_dimension, max_value=maximum_dimension
+            )
+        )
+    )
+    _ensure_dimension_is_even_and_in_bounds(
+        maximum_dimension, minimum_dimension, even_dimension
+    )
+    return even_dimension
```
[the remark this answers is not in the document]
Comment (BjoernLudwigPTB) [reply test/conftest.py:362]: Resolved by 7bf7dbd2.

File: test/test_tools.py (hunk 1 above)
Comment (BjoernLudwigPTB) [reply test/test_tools.py:279]: Resolved by 173481a4 and 48ab4a73

File: test/test_tools.py (hunk 2 above)
Comment (BjoernLudwigPTB) [reply test/test_tools.py]: Resolved by 3cb47289

File: test/conftest.py [hunk 9]
```diff
@@ -387,7 +476,9 @@ def hypothesis_two_to_the_k(
     draw: Callable, min_k: Optional[int] = None, max_k: Optional[int] = None
```
[the remark this answers is not in the document]
Comment (BjoernLudwigPTB) [reply test/conftest.py]: Resolved by b0319b7

**Pull request merged**